### PR TITLE
KerasNLP 0.4 API documentation

### PR DESCRIPTION
**Pull request by mattdangerw**
Not ready to merge quite yet, but opening a draft PR so people can comment.

### CODE REVIEW

File: scripts/nlp_api_master.py [hunk 1]
```diff
@@ -69,82 +245,115 @@
...
         },
-        {
-            "path": "mlm_mask_generator",
-            "title": "MLMMaskGenerator layer",
-            "generate": ["keras_nlp.layers.MLMMaskGenerator"],
-        },
         {
             "path": "mlm_head",
-            "title": "MLMHead layer",
+            "title": "MLMHead",
```
Comment (fchollet) [inline scripts/nlp_api_master.py:354]: We should be consistent with core Keras docs, which currently have "layer" in the page title for all layers, e.g. https://keras.io/api/layers/normalization_layers/batch_normalization/

File: scripts/nlp_api_master.py [hunk 2]
```diff
@@ -156,19 +365,30 @@
     "children": [
         {
             "path": "perplexity",
-            "title": "Perplexity metric",
+            "title": "Perplexity",
```
Comment (fchollet) [inline scripts/nlp_api_master.py:368]: Likewise it's likely more consistent to keep the object type in the title

File: templates/api/keras_nlp/index.md [hunk 3]
```diff
@@ -1,8 +1,9 @@
 # KerasNLP
 
-KerasNLP is a toolbox of modular building blocks (layers, metrics, etc.) that
-NLP engineers can leverage to develop
-production-grade, state-of-the-art training and inference pipelines for common NLP workflows.
+KerasNLP is a toolbox of modular building blocks ranging from pre-trained
```
Comment (fchollet) [inline templates/api/keras_nlp/index.md:3]: We tend to use `pretrained` as a single word across the docs

File: templates/api/keras_nlp/index.md [hunk 4]
```diff
@@ -1,8 +1,9 @@
...
+state-of-the-art models, to low-level transformer encoder layers. For an
```
Comment (fchollet) [inline templates/api/keras_nlp/index.md:4]: Capitalize Transformer Encoder

File: templates/api/keras_nlp/models/index.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+1) Through the `from_preset()` constructor, which instantiates an object with
+a pre-trained configurations, vocabularies, and (optionally) weights. Available
+preset IDs are listed on this page.
+
+2) Through custom configuration controlled by the user. To do this, simply
+pass the desired configuration parameters to the default constructors of the
+symbols documented below.
+
+## Backbone presets
```
Comment (fchollet) [inline templates/api/keras_nlp/models/index.md:14]: It may be useful to start with a picture that shows the relationship between classifier, backbone, preprocessor, and tokenizer, since they are the key mental models users need

File: templates/api/keras_nlp/models/index.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+| xlm_roberta_base             | [XLM-RoBERTa](xlm_roberta) | 277M        | 12-layer XLM-RoBERTa model where case is maintained. Trained on CommonCrawl in 100 languages. |
+| xlm_roberta_large            | [XLM-RoBERTa](xlm_roberta) | 558M        | 24-layer XLM-RoBERTa model where case is maintained. Trained on CommonCrawl in 100 languages. |
+| distil_bert_base_en_uncased  | [DistilBert](distil_bert)  | 66M         | 6-layer DistilBERT model where all input is lowercased. Trained on English Wikipedia + BooksCorpus. |
+| distil_bert_base_en_cased    | [DistilBert](distil_bert)  | 65M         | 6-layer DistilBERT model where case is maintained. Trained on English Wikipedia + BooksCorpus. |
+| distil_bert_base_multi_cased | [DistilBert](distil_bert)  | 134M        | 6-layer DistilBERT model where case is maintained. Trained on Wikipedias of 104 languages. |
+
+## Classification presets
+
+The following preset ID correspond to a configuration, weights and vocabulary
+for a model **classifier**. These models are inference ready, but can be further
```
Comment (fchollet) [inline templates/api/keras_nlp/models/index.md:49]: inference-ready

File: templates/api/keras_nlp/models/index.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+fine-tuned if desired.
+
+IDs below can be used with the `from_preset()` constructor for classifier models
+and preprocessing layers.
+For example, a BERT classifier preset can be used with
+`keras_nlp.models.BertTokenizer`,
+`keras_nlp.models.BertPreprocessor`, or
+`keras_nlp.models.BertClassifier`.
```
Comment (fchollet) [inline templates/api/keras_nlp/models/index.md:57]: Might be useful to include an explicit code snippet like

```python
classifier = BertClassifier.from_preset("bert_tiny_en_uncased")
backbone = BertBackbone.from_preset("bert_tiny_en_uncased")
tokenizer = BertTokenizer.from_preset("bert_tiny_en_uncased")
preprocessor = BertPreprocessor.from_preset("bert_tiny_en_uncased")
```

File: templates/api/keras_nlp/models/index.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
+# KerasNLP Models
+
+KerasNLP contains end-to-end implementations of popular model
+architectures. These models can be created in two ways:
+
+1) Through the `from_preset()` constructor, which instantiates an object with
```
Comment (fchollet) [inline templates/api/keras_nlp/models/index.md:6]: Show a 1-line code snippet to illustrate

File: templates/api/keras_nlp/models/index.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+a pre-trained configurations, vocabularies, and (optionally) weights. Available
+preset IDs are listed on this page.
+
+2) Through custom configuration controlled by the user. To do this, simply
```
Comment (fchollet) [inline templates/api/keras_nlp/models/index.md:10]: Show a short code snippet to illustrate this for e.g. a classifier, a backbone and a preprocessor

**Review by fchollet — COMMENTED** [review]
Looks great, thank you for the PR!

File: templates/api/keras_nlp/models/index.md (hunk 5 above)
Comment (mattdangerw) [reply templates/api/keras_nlp/models/index.md:14]: I think for now, I will just link to the getting started guide, which has this.

File: templates/api/keras_nlp/models/index.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+The following preset ID correspond to a configuration, weights and vocabulary
+for a model **classifier**. These models are inference ready, but can be further
+fine-tuned if desired.
+
+IDs below can be used with the `from_preset()` constructor for classifier models
+and preprocessing layers.
+
+```python
+backbone = keras_nlp.models.BertBackbone.from_preset("bert_tiny_en_uncased_sst2")
```
Comment (mattdangerw) [inline templates/api/keras_nlp/models/index.md:88]: this should be classifier!!!

**Comment by mattdangerw** [comment]
This won't pass any CI until we actually push 0.4 to pypi, but this should be ready otherwise!

**Review by fchollet — APPROVED** [review]
LGTM -- should I merge (which would break CI), or do we wait?

**Comment by mattdangerw** [comment]
> LGTM -- should I merge (which would break CI), or do we wait?

About to do the actual pypi release with a few very minor changes, the will update this PR (hopefully no breakages then). Will ping when this is ready!

**Comment by mattdangerw** [comment]
OK! This should be all set, and the pip release is out, so this will not cause breakages (it will fix them now hopefully as the API docs are now out of date).

Ready to merge whenever.
